File: services/engine/internal/auth/rbac.go
package auth

import (
	"strings"
)

// Canonical platform roles.
const (
	RoleAdmin    = "admin"
	RoleTrader   = "trader"
	RoleRisk     = "risk"
	RoleOperator = "operator"
	RoleViewer   = "viewer"
)

// CanAccess returns true if any session role matches an allowed role (after normalization).
func CanAccess(sessionRoles []string, allowed []string) bool {
	if len(allowed) == 0 {
		return true
	}
	for _, sr := range sessionRoles {
		u := NormalizeRole(sr)
		for _, a := range allowed {
			if u == NormalizeRole(a) {
				return true
			}
		}
	}
	return false
}

// NormalizeRole lowercases and trims; unknown values fall back to viewer.
func NormalizeRole(role string) string {
	r := strings.ToLower(strings.TrimSpace(role))
	switch r {
	case RoleAdmin, RoleTrader, RoleRisk, RoleOperator, RoleViewer:
		return r
	default:
		return RoleViewer
	}
}

// StepUpRequired marks privileged actions that require an additional step-up token header.
func StepUpRequired(action string) bool {
	a := strings.ToLower(strings.TrimSpace(action))
	switch a {
	case "withdraw", "pause-trading", "rotate-keys", "policy-change", "user-delete", "sudo":
		return true
	default:
		return false
	}
}

File: apps/dashboard/lib/rbac.ts
export type Role = "admin" | "trader" | "risk" | "operator" | "viewer";

export function normalizeRole(role: string | undefined | null): Role {
  const r = (role ?? "viewer").toLowerCase().trim();
  if (r === "admin" || r === "trader" || r === "risk" || r === "operator" || r === "viewer") {
    return r;
  }
  return "viewer";
}

export function canAccess(sessionRoles: string[], allowed: Role[]): boolean {
  if (allowed.length === 0) return true;
  const set = new Set(allowed.map((a) => a.toLowerCase()));
  return sessionRoles.some((sr) => set.has(normalizeRole(sr)));
}

export function stepUpRequired(action: string): boolean {
  const a = action.toLowerCase().trim();
  return ["withdraw", "pause-trading", "rotate-keys", "policy-change", "user-delete"].includes(a);
}
